sdk/src/main/java/com/qonversion/android/sdk/QProductCenterManager.kt

```diff
@@ -103,6 +103,44 @@ class QProductCenterManager internal constructor(
         oldProductId: String?,
         @BillingFlowParams.ProrationMode prorationMode: Int?,
         callback: QonversionPermissionsCallback
+    ) {
+        if (launchError != null) {
+            val launchCallback = getLaunchCallback(object: QonversionLaunchCallback {
+                override fun onSuccess(launchResult: QLaunchResult) {
+                    if (isProductsLoaded && !isProductsLoadingFailed) {
+                        processPurchase(context, id, oldProductId, prorationMode, callback)
+                    } else {
+                        val productsCallback = object : QonversionProductsCallback {
+                            override fun onSuccess(products: Map<String, QProduct>) {
+                                processPurchase(context, id, oldProductId, prorationMode, callback)
+                            }
+
+                            override fun onError(error: QonversionError) {
+                                callback.onError(error)
+                            }
+                        }
+
+                        productsCallbacks.add(productsCallback)
+                    }
```

Comment:
productsCallback is not so big, so you can define it as an inside argument:
 productsCallbacks.add(object : QonversionProductsCallback {
            override fun onSuccess(products: Map<String, QProduct>) {
                processPurchase(context, id, oldProductId, prorationMode, callback)
            }

            override fun onError(error: QonversionError) {
                callback.onError(error)
            }
        })